synapse/storage/databases/main/events_worker.py

```diff
@@ -2111,7 +2111,14 @@ async def get_event_id_for_timestamp(
                 AND room_id = ?
                 /* Make sure event is not rejected */
                 AND rejections.event_id IS NULL
-            ORDER BY origin_server_ts %s
+            /**
+             * First sort by the message timestamp. If the message timestamps are the
+             * same, we want the message that logically comes "next" (before/after
+             * the given timestamp) based on the DAG and its topological order (`depth`).
+             * Finally, we can tie-break based on when it was received on the server
+             * (`stream_ordering`).
+             */
+            ORDER BY origin_server_ts %s, depth %s, stream_ordering %s
```

Comment:
> when it was received on the server (`stream_ordering`).

There is a received_ts column for this (though there may very well be tie-breaks just like the origin timestamp).

origin_server_ts and depth are untrusted fields controlled by the sender's HS. I'm not sure if that is a concern here though.

Reply:
I can't think of a better way to do this without ordering by `depth`. There is a `topological_ordering` column, but that's just identical to `depth`.

We're already trusting `origin_server_ts` to be reasonable for this feature, so I suppose it's not much of a stretch to trust `depth` as well.

Reply:
> origin_server_ts and depth are untrusted fields controlled by the sender's HS. I'm not sure if that is a concern here though.

> We're already trusting `origin_server_ts` to be reasonable for this feature, so I suppose it's not much of a stretch to trust `depth` as well.

👍 exactly this